### ISSUE TITLE
`ImportOrder.separated=false` not respected

### ISSUE THREAD

**Issue by jonmbake** [checkstyle/checkstyle#5476]
`/var/tmp $ javac ImportOrderTest.java`:

```
```

`/var/tmp $ cat ImportOrderTest.java`:

```
import java.io.Serializable;

import java.util.List;

/**
 * Test for CheckStyle's ImportOrderCheck w/ spaces.
 */
public class ImportOrderTest {
}
```

`/var/tmp $ cat config.xml`:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ImportOrder">
            <property name="separated" value="false"/>
        </module>
    </module>
</module>
```

`/var/tmp $ java -Duser.language=en -Duser.country=US -jar checkstyle-6.4.1-all.jar -c config.xml ImportOrderTest.java`:

```
Starting audit...
Audit done.
```

`/var/tmp $ java -Duser.language=en -Duser.country=US -jar checkstyle-8.7-all.jar -c config.xml ImportOrderTest.java`:

```
Starting audit...
[ERROR] /Users/jonbake/Code/git/checkstyle-import-order-test/src/main/java/ImportOrderTest.java:3: Extra separation in import group before 'java.util.List' [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

Originally referenced here: https://github.com/checkstyle/checkstyle/issues/2143#issuecomment-359028352.

Spaces between imports should not give an error when `ImportOrder.separated` is set to false.

This is a regression. 6.4.1 does not give an error; 8.7 does. Complete code here: https://github.com/jonmbake/checkstyle-import-order-test. To see error, run `mvn validate`.

--------------

**Comment by varunu28** [comment]
@romani I would like to take up this issue

**Comment by romani** [comment]
@varunu28 , there are already bunch active PRs for this Check, around the same problem.

please get confirmation that you will not make problems for @djydewang .
@djydewang , please share your thoughts on this issue.

**Comment by djydewang** [comment]
@romani One thing needs be confirmed. 

> Imo, the only thing that has to be done on the user side, is, to add a flag (true/false) if it's allowed to separate these groups internally. https://github.com/checkstyle/checkstyle/issues/2143#issuecomment-138673671

Should CS check separator between groups when `separator=false`?
CS before 8.3 vesion would throw nothing about separator when `separator=false`.
For example
```
import java.util.Set;

import org.*;
```

```
import java.util.Set;
import org.*;
```
They are both OK for the configuration:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ImportOrder">
            <property name="groups" value="java, org"/>
            <property name="separated" value="false"/>
        </module>
    </module>
</module>
```

**Comment by romani** [comment]
>Should CS check separator between groups when separator=false?

As checkstyle demand behavior, and all other configuration properties demand strictly follow rules - it is ok to make violation on this.

**Comment by djydewang** [comment]
@romani I thought it over again.
If `separator=false` means that there should be no separator between groups and allow user separate groups internally, then the following valid imports format is strange


```
import java.io.Serializable;

import java.io.*;
import org.*;
```

**Comment by jonmbake** [comment]
My 2 cents:

I don't think the default _ImportOrder_ check, e.g.:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="ImportOrder"/>
    </module>
</module>
```
Should do any checking for line space (separation) between imports. Seems like separator spacing should be an option that that has to be explicitly set.

Is there a way to use the `ImportOrder` check and not have it error for separator spacing? I just want the imports ordered; I don't care if there is any spacing between imports, whether that is a group or not.

**Comment by RohanNagar** [comment]
I agree with @jonmbake. Since 8.3 all of my import checks have been broken. Is there no way to get the behavior of `ImportOrder` before 8.3? I don't care about spacing either, I just want alphabetical order.

**Comment by djydewang** [comment]
@romani What's your opinion?

**Comment by jonmbake** [comment]
@djydewang @romani I am willing to submit a PR to fix this issue if you guys decide it is the behavior you want.

**Comment by djydewang** [comment]
@romani ping

**Comment by RohanNagar** [comment]
For those of you wanting to get back to similar behavior prior to v8.3, you can use `CustomImportOrder`.

Replace
```
<module name="ImportOrder">
  ...
</module>
```

with

```xml
<module name="CustomImportOrder">
  <property name="sortImportsInGroupAlphabetically" value="true"/>
  <property name="separateLineBetweenGroups" value="true"/>
  <property name="customImportOrderRules" value="THIRD_PARTY_PACKAGE###STATIC"/>
</module>
```

and see options for configuration [here](http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.html).

**Comment by romani** [comment]
@djydewang ,

>So ImportOrder.separator=false should allow user separate groups internally.
Should CS check separator between groups when separator=false?

By logic of Check:
As we say groups are separated .... it mean we expect empty line between groups for sure.
So "group" is smth that is separated by empty line.
it mean that any extra empty line inside group is a sign of subgroups, ... so groups.

All that is separated by empty lines is a group.

===============
@jonmbake

> Should do any checking for line space (separation) between imports

Please read my definition of "group" above.
Please share with us details on why you would like to keep empty lines in groups.
Your cases a might be bit artificial, so please provide real user story of this Check usage.
Checkstyle is designed to demand format/style/good-practices so we not ok if user just lazy to cleanup the code.

**Comment by RohanNagar** [comment]
@romani I was just trying to follow Google's Style Guide and didn't realize they had updated to use `CustomImportOrder` rather than just `ImportOrder`. In general I think that style checking is very subjective and so there should be as many options as possible within checkstyle to allow for customization. Since I was able to get this feature through `CustomImportOrder`, I am satisfied.

**Comment by romani** [comment]
File https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml is maintained by our team, not by Google , so misinterpretations are possible. As far as I remember, reason was not in skipping empty lines. Probably one day such fix( no extra empty lines) come to CustomImportOrder too.

>In general I think that style checking is very subjective and so there should be as many options as possible within checkstyle to allow for customization

Kind of yes, but we try to validate that reason is deserve to be part of main library and maintained, so we spent special attention to design. We can not satisfy all but we try to  be reasonably open  to suggestions and requests.If no good reasons, we reject proposal and author should make custom Check and tune it to his needs.

**Comment by romani** [comment]
@jonmbake , please respond to support your issue.

if no respond till end of April I will close the issue.

**Comment by noelo-cohelo** [comment]
Hi guys :)

I have the exact problem as in first post in this topic - with the file **ImportOrderTest** I can't set up Checkstyle so it doesn't complain (seems like separated=false/true doesn't make any difference).

Unfortunately, you mixed here some stuff about _separating groups internally_ (which I'm not interested in), although this was supposed to be another bug from what I understand.

Now I'm a bit confused, but still I would love to have it working. I really don't care if this means that I **can** have separators between lines, or I **must** - the second seems to be better option if this tag is present. But now I don't see any possibility to set **ImportOrder** so.

**Comment by antoinetran** [comment]
> 
> 
> @jonmbake , please respond to support your issue.
> 
> if no respond till end of April I will close the issue.

Hi, I just want to add my two cents here since our project set `separated=false` for whatever reason, and share the Checkstyle configuration to our subcontractors. Now each time we change a shared configuration, it must undergo a review and there can be financial impacts (especially for a change from our initiative).
Thus, making a breaking change like this without a retro-compatibility option is not great for us, because we either keep our old CheckStyle version or pay a cost of migration. We need to spend time in this CustomImportOrder to find a similar (exact?) behavior than before and in worst case pay our subcontractor for them to update the CheckStyle policy.

Thanks @RohanNagar for the example of CustomImportOrder!

**Comment by jonmbake** [comment]
I used the workaround w/ the CustomImportOrder so this is no longer an issue for me. I still hold my opinion in https://github.com/checkstyle/checkstyle/issues/5476#issuecomment-359971988, but since there is a workaround, it is less of an issue IMO.

**Comment by antoinetran** [comment]
Ok, after I looked at how to adapt CustomImportOrder to reproduce ImportOrder module... I discovered that this code:
```
		<module name="ImportOrder">
			<metadata name="net.sf.eclipsecs.core.comment" value="EUM-JCRule-680: Import order check" />
			<property name="ordered" value="false" />
			<property name="separated" value="false" />
		</module>
```
does absolutely nothing so I just removed it! And the workaround does not seem that hard to get so to my mind, ok for only the workaround. As for me, I removed this module.

**Comment by serianox** [comment]
Hi all,

This is impacting us too. Though I haven't tested with the latest version yet, I assume from the current discussion that this issue is still active.

We have import groups that may contains comments for business reasons. The comments are improperly flagged as extra separators, whatever the value of _separated_.

Currently, my best option is to deactivate _ImportOrder_ completely. I'm trying to figure out of CustomImportOrder can be used to somehow emulate the behavior.

**Comment by romani** [comment]
>We have import groups that may contains comments for business reasons.

Can you share them? To let us know why people put comment in imports. We thought that import management is purely done by IDE, not a human.

**Comment by serianox** [comment]
Unfortunately, I can't share as much. The furthest I can tell is that we annotate some of the imports to circumvent the limitations of the Java language in the Java Card Runtime Environment.

Actually, the perfect fix - for us - would be not to return an error if a group contains comments. But as you mentioned, this is not the expected behavior in modern Java dev where the IDE manages the import statements for you.

Here, I just wanted to use _separated=false_ to relax the check, hoping that not enforcing no spaces inside groups would also allow for comments. I would not enforce spaces between groups, but at least I'll enforce sorting of groups and won't get those false positives. But with this bug, I'm (probably) getting false positives from those comments.

Finally, I'm just post-processing the Checkstyle xml output to remove the false positive in CI, but that's not really optimal. 😺

**Comment by romani** [comment]
I hope you aware of our filters https://checkstyle.org/config_filters.html so you can filter out by check and message all violations that you do not like.

**Comment by serianox** [comment]
I... uh... no. 🙀

For those reading the thread after, here's how I suppressed the messages (takes care of the locale)
```xml
<module name="SuppressionSingleFilter">
  <property name="message" value="HIDE ME"/>
</module>

<module name="ImportOrder">
  <message key="import.groups.separated.internally" value="HIDE ME"/>
```
